ifermi/surface.py

```diff
@@ -315,15 +324,37 @@ def all_vertices_faces(
             spins = [spins]
 
         vertices_faces = []
-        for spin in spins:
-            for isosurface in self.isosurfaces[spin]:
-                vertices_faces.append((isosurface.vertices, isosurface.faces))
+        if band_index is None:
+            for spin in spins:
+                for isosurface in self.isosurfaces[spin]:
+                    vertices_faces.append((isosurface.vertices, isosurface.faces))
+        elif isinstance(band_index, int):
+            for spin in spins:
+                vertices_faces.append((self.isosurfaces[spin][band_index].vertices,
+                                       self.isosurfaces[spin][band_index].faces))
+        elif isinstance(band_index, list):
+            for spin in spins:
+                for i in band_index:
+                    vertices_faces.append((self.isosurfaces[spin][i].vertices,
+                                           self.isosurfaces[spin][i].faces))
+        elif isinstance(band_index, dict):
+            for spin in self.spins:
+                if spin in band_index:
+                    if isinstance(band_index[spin], int):
+                        vertices_faces.append((self.isosurfaces[spin][band_index[spin]].vertices,
+                                               self.isosurfaces[spin][band_index[spin]].faces))
+                    elif isinstance(band_index[spin], List):
+                        for i in band_index[spin]:
+                            vertices_faces.append((self.isosurfaces[spin][i].vertices,
+                                                   self.isosurfaces[spin][i].faces))
+
```

Comment:
This is really nice and exactly what I had in mind. I think it could be simplified in the case where band_index is not None. I think if you can get the band_index into a common format then you should only need one additional else statement. I.e., first turn band index into a dict of lists, and then you can just handle that case!

So something along the lines of (note I didn't test this):
```suggestion
        if not isinstance(band_index, dict):
            band_index = {spin: band_index for spin in spins}

        for spin, spin_index in band_index.items():
            if isinstance(spin_index, int):
                band_index[spin] = [spin_index]

        if band_index is None:
            for spin in spins:
                for isosurface in self.isosurfaces[spin]:
                    vertices_faces.append((isosurface.vertices, isosurface.faces))
        else:
            for spin in spins:
                if spin in band_index:
                    for i in band_index[spin]:
                        vertices_faces.append((self.isosurfaces[spin][i].vertices,
                                               self.isosurfaces[spin][i].faces))
```

You should be able to use this to simplify all_properties in the same way.

Reply:
In fact, you could go even further, and handle the case where band_index is None. I.e,

```suggestion
        if band_index is None:
            band_index = {spin: list(range(len(spin_surfaces))
                          for spin, spin_surfaces in self.isosurfaces.items())}

        if not isinstance(band_index, dict):
            band_index = {spin: band_index for spin in spins}

        for spin, spin_index in band_index.items():
            if isinstance(spin_index, int):
                band_index[spin] = [spin_index]

        for spin in spins:
            if spin in band_index:
                for i in band_index[spin]:
                    vertices_faces.append((self.isosurfaces[spin][i].vertices,
                                           self.isosurfaces[spin][i].faces))
```

Reply:
Tested this out and it works except the ) bracket on the third line needs to come after `spin_surfaces`. I've put it into the code plus done the same for the slices. It seems to be working from tests!